python/ray/tests/test_gcs_fault_tolerance.py

```diff
@@ -647,6 +648,93 @@ def pid(self):
         ray.get_actor("A")
 
 
+@pytest.fixture
+def redis_replicas(monkeypatch):
+    monkeypatch.setenv("TEST_EXTERNAL_REDIS_REPLICAS", "3")
+
+
+@pytest.mark.parametrize(
+    "ray_start_cluster_head_with_external_redis",
+    [
+        generate_system_config_map(
+            gcs_failover_worker_reconnect_timeout=20,
+            gcs_rpc_server_reconnect_timeout_s=60,
+            gcs_server_request_timeout_seconds=10,
+            redis_db_connect_retries=50,
+        )
+    ],
+    indirect=True,
+)
+def test_redis_failureover(redis_replicas, ray_start_cluster_head_with_external_redis):
+    cluster = ray_start_cluster_head_with_external_redis
+    import redis
+
+    redis_addr = os.environ.get("RAY_REDIS_ADDRESS")
+    ip, port = redis_addr.split(":")
+    cli = redis.Redis(ip, port)
+    nodes = cli.cluster("nodes")
+    leader_cli = None
+    follower_cli = []
+    for n in nodes:
+        addr, _ = n.split("@")
+        ip, port = addr.split(":")
+        cli = redis.Redis(ip, port)
+        meta = nodes[n]
+        flags = meta["flags"].split(",")
+        if "master" in flags:
+            leader_cli = cli
+        else:
+            follower_cli.append(cli)
+
+    leader_pid = leader_cli.info()["process_id"]
+
+    @ray.remote(max_restarts=-1)
+    class Counter:
+        def r(self, v):
+            return v
+
+    c = Counter.options(name="c", lifetime="detached").remote()
+    r = ray.get(c.r.remote(10))
+    assert r == 10
+
+    head_node = cluster.head_node
+    gcs_server_process = head_node.all_processes["gcs_server"][0].process
+    gcs_server_pid = gcs_server_process.pid
+
+    # Wait until all data is updated in the replica
+    sleep(5)
+
+    # Now kill pid
+    leader_process = psutil.Process(pid=leader_pid)
+    leader_process.kill()
+
+    print(">>> Waiting gcs server to exit", gcs_server_pid)
+    wait_for_pid_to_exit(gcs_server_pid, 1000)
+
+    print("GCS killed")
+    follower_cli[0].cluster("failover", "takeover")
```

Comment:
this is essentially emulating what a managed redis solution should do, right?

Reply:
have we verified that memorydb follows the same behavior?

Reply:
No, I'm not and I'm not quite sure how to do that with MemoryDB since it's not managed by me.

I think you mentioned MemoryDB will manage this for us right? Otherwise, I think we shouldn't use MemoryDB.